quality-debt: .agents/scripts/rosetta-audit-helper.sh — PR #1184 review feedback (high)

## Unactioned Review Feedback

**Source PR**: #1184
**File**: `.agents/scripts/rosetta-audit-helper.sh`
**Reviewers**: coderabbit, gemini
**Findings**: 5
**Max severity**: high

---

### MEDIUM: coderabbit (coderabbitai[bot])
**File**: `.agents/scripts/rosetta-audit-helper.sh:57`
_⚠️ Potential issue_ | _🟡 Minor_

**`set -e` interaction with bare `&& return $?` pattern is fragile.**

Both `is_apple_silicon` and `has_dual_brew` use `[[ ]] && [[ ]]; return $?`. Under `set -e`, the command after the final `&&` is **not** exempt from errexit. If the second test fails, `set -e` would terminate the script — unless the function is called inside an `if`/`||` context (which suppresses errexit for the whole function body).

Currently all call sites use `if ! is_apple_silicon; then`, so this is safe **today**. But a future direct call (e.g., `is_apple_silicon && do_something`) would cause an unexpected script abort on Intel Macs.



<details>
<summary>🛡️ Defensive rewrite</summary>

```diff
 is_apple_silicon() {
-    [[ "$(uname)" == "Darwin" ]] && [[ "$(uname -m)" == "arm64" ]]
-    return $?
+    if [[ "$(uname)" == "Darwin" ]] && [[ "$(uname -m)" == "arm64" ]]; then
+        return 0
+    fi
+    return 1
 }
 
 has_dual_brew() {
-    [[ -x "$X86_BREW" ]] && [[ -x "$ARM_BREW" ]]
-    return $?
+    if [[ -x "$X86_BREW" ]] && [[ -x "$ARM_BREW" ]]; then
+        return 0
+    fi
+    return 1
 }
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In @.agents/scripts/rosetta-audit-helper.sh around lines 44 - 54, The functions
is_apple_silicon and has_dual_brew rely on the fragile pattern "[[ ]] && [[ ]];
return $?" which can cause unexpected exits under set -e; change each to an
explicit boolean return path so the function body never lets a failing test
trigger errexit — e.g., evaluate the compound condition and then return 0 on
success and return 1 on failure (or use a guarded "if ...; then return 0; else
return 1; fi" around the existing tests) so callers like "is_apple_silicon &&
do_something" won't cause the script to abort unexpectedly.
```

</details>

<!-- fingerprinting:phantom:medusa:phoenix -->

<!-- This is an auto-generated comment by CodeRabbit -->

✅ Addressed in commit 3385c5b

[View comment](https://github.com/marcusquinn/aidevops/pull/1184#discussion_r2796207000)
---

### MEDIUM: coderabbit (coderabbitai[bot])
**File**: `.agents/scripts/rosetta-audit-helper.sh:165`
_⚠️ Potential issue_ | _🟡 Minor_

**Temp files not protected by a trap — will leak on unexpected exit.**

`cmd_scan` creates two temp files via `mktemp` and relies on manual `rm -f` at lines 201 and 234. If the script exits between creation and cleanup (signal, `set -e` triggering inside `categorise_packages`, etc.), the files persist in `/tmp`.

The codebase already has `_save_cleanup_scope` / `push_cleanup` (used in `check_positional_parameters` at line 137). Using the same pattern here would be consistent and safer.



<details>
<summary>🛡️ Use existing cleanup pattern</summary>

```diff
     migratable_file=$(mktemp "${TMPDIR:-/tmp}/rosetta-migrate.XXXXXX")
     non_migratable_file=$(mktemp "${TMPDIR:-/tmp}/rosetta-nomigrate.XXXXXX")
-    _SCAN_CLEANUP_FILES="$migratable_file $non_migratable_file"
+    _save_cleanup_scope; trap '_run_cleanups' RETURN
+    push_cleanup "rm -f '$migratable_file' '$non_migratable_file'"
```

Then remove the manual `rm -f $_SCAN_CLEANUP_FILES` calls at lines 201 and 234.
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In @.agents/scripts/rosetta-audit-helper.sh around lines 156 - 161, cmd_scan
creates temporary files (migratable_file, non_migratable_file) and stores them
in _SCAN_CLEANUP_FILES but doesn’t register them with the existing cleanup
mechanism, so they can leak on unexpected exit; modify cmd_scan to call
_save_cleanup_scope and push_cleanup (the same pattern used by
check_positional_parameters) immediately after creating migratable_file and
non_migratable_file so the files are removed on exit, and then delete the manual
rm -f $_SCAN_CLEANUP_FILES calls (leave _SCAN_CLEANUP_FILES assignment intact so
other code can reference it).
```

</details>

<!-- fingerprinting:phantom:medusa:phoenix -->

<!-- This is an auto-generated comment by CodeRabbit -->

✅ Addressed in commit 3385c5b

[View comment](https://github.com/marcusquinn/aidevops/pull/1184#discussion_r2796207003)
---

### HIGH: coderabbit (coderabbitai[bot])
**File**: `.agents/scripts/rosetta-audit-helper.sh:371`
_⚠️ Potential issue_ | _🟡 Minor_

**`cmd_migrate` always returns 0, even with failures — callers can't detect partial failure.**

When `failed > 0`, the function prints a warning but still `return 0`. If `setup.sh` or any caller checks the exit code to decide next steps, failures would be silently ignored.



<details>
<summary>🐛 Proposed fix</summary>

```diff
     # Invalidate cache
     rm -f "$SCAN_CACHE" 2>/dev/null || true
 
-    return 0
+    if [[ "$failed" -gt 0 ]]; then
+        return 1
+    fi
+    return 0
 }
```
</details>

<!-- suggestion_start -->

<details>
<summary>📝 Committable suggestion</summary>

> ‼️ **IMPORTANT**
> Carefully review the code before committing. Ensure that it accurately replaces the highlighted code, contains no missing lines, and has no issues with indentation. Thoroughly test & benchmark the code to ensure it meets the requirements.

```suggestion
    echo "  Failed:                 $failed"
    echo ""

    if [[ "$dry_run" = true ]]; then
        print_info "Run without --dry-run to execute migration"
    elif [[ "$failed" -eq 0 ]]; then
        print_success "Migration complete"

        # Check if x86 brew is now empty
        local remaining
        remaining=$("$X86_BREW" list --formula 2>/dev/null | wc -l | tr -d ' ')
        if [[ "$remaining" -eq 0 ]]; then
            echo ""
            print_info "x86 Homebrew is now empty. You can remove it:"
            echo "  sudo rm -rf /usr/local/Homebrew"
            echo "  sudo rm -rf /usr/local/Cellar"
            echo "  sudo rm -rf /usr/local/bin/brew"
        else
            echo ""
            print_info "$remaining x86 packages remain (no ARM formula or have dependents)"
        fi
    else
        print_warning "Migration completed with $failed failures"
    fi

    # Invalidate cache
    rm -f "$SCAN_CACHE" 2>/dev/null || true

    if [[ "$failed" -gt 0 ]]; then
        return 1
    fi
    return 0
}
```

</details>

<!-- suggestion_end -->

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In @.agents/scripts/rosetta-audit-helper.sh around lines 339 - 368, The
cmd_migrate function currently always returns 0; update its exit behavior so
callers can detect failures by returning a non-zero status when failures
occurred: after the block that checks "$failed" and prints the warning (the
branch that prints "Migration completed with $failed failures"), change the
final return to return a non-zero value (e.g., return $failed or return 1) while
keeping dry-run and success paths returning 0; locate the cmd_migrate function
and the "$failed" variable usage to implement this change so setup.sh and other
callers can react to partial failures.
```

</details>

<!-- fingerprinting:phantom:medusa:phoenix -->

<!-- This is an auto-generated comment by CodeRabbit -->

✅ Addressed in commit 3385c5b

[View comment](https://github.com/marcusquinn/aidevops/pull/1184#discussion_r2796207004)
---

### HIGH: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/rosetta-audit-helper.sh:160`
![high](https://www.gstatic.com/codereviewagent/high-priority.svg)

The script's temporary file handling can be improved for robustness and to adhere to the style guide.

1.  **Guaranteed Cleanup**: A `trap` should be used to ensure temporary files are removed even if the script exits unexpectedly. This is required by style guide rule #33.
2.  **Robust File List**: The `_SCAN_CLEANUP_FILES` global variable should be an array to be robust against filenames with special characters and avoid `shellcheck disable=SC2086`.

I recommend defining `_SCAN_CLEANUP_FILES` as an array and setting an `EXIT` trap at the top of the script.

**At top of script (e.g., after `set -euo pipefail`):**
```bash
_SCAN_CLEANUP_FILES=()
trap 'rm -f "${_SCAN_CLEANUP_FILES[@]}"' EXIT
```

Then, this block can be updated to populate the array. The explicit `rm` calls at the end of `cmd_scan` can then be removed.

```suggestion
    local migratable_file non_migratable_file
    migratable_file=$(mktemp "${TMPDIR:-/tmp}/rosetta-migrate.XXXXXX")
    non_migratable_file=$(mktemp "${TMPDIR:-/tmp}/rosetta-nomigrate.XXXXXX")
    _SCAN_CLEANUP_FILES=("$migratable_file" "$non_migratable_file")
```

<details>
<summary>References</summary>

1. Temp files must have `trap` cleanup (RETURN or EXIT). The current implementation creates temp files but does not guarantee their cleanup if the script exits unexpectedly. <sup>([link](https://github.com/marcusquinn/aidevops/blob/main/.gemini/styleguide.md))</sup>
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/1184#discussion_r2796210399)
---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/rosetta-audit-helper.sh:279`
![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)

Redirecting stderr to `/dev/null` suppresses potentially useful error messages from `brew uninstall`. This makes debugging failures difficult and violates style guide rule #50, which states that `2>/dev/null` should only be used when redirecting to log files, not for blanket suppression. Since the command is inside an `if` statement, removing the redirection will print the error without exiting the script, providing better diagnostic information on failure.

```suggestion
                if "$X86_BREW" uninstall "$pkg"; then
```

<details>
<summary>References</summary>

1. `2>/dev/null` is acceptable ONLY when redirecting to log files, not blanket suppression. This code uses it to suppress errors from brew commands, which hinders debugging. <sup>([link](https://github.com/marcusquinn/aidevops/blob/main/.gemini/styleguide.md))</sup>
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/1184#discussion_r2796210409)
---

---
_Auto-generated by `quality-feedback-helper.sh scan-merged`. Review each finding and either fix the code or dismiss with a reason._

--- a/.agents/scripts/rosetta-audit-helper.sh
+++ b/.agents/scripts/rosetta-audit-helper.sh
@@ -23,6 +23,10 @@ set -euo pipefail
 SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)" || exit
 source "${SCRIPT_DIR}/shared-constants.sh"
 
+# Temp file registry — populated by cmd_scan; cleaned up on EXIT (style guide rule #33)
+_SCAN_CLEANUP_FILES=()
+trap 'rm -f "${_SCAN_CLEANUP_FILES[@]}"' EXIT
+
 # =============================================================================
 # Constants
 # =============================================================================
@@ -159,6 +163,7 @@ cmd_scan() {
 	local migratable_file non_migratable_file
 	migratable_file=$(mktemp "${TMPDIR:-/tmp}/rosetta-migrate.XXXXXX")
 	non_migratable_file=$(mktemp "${TMPDIR:-/tmp}/rosetta-nomigrate.XXXXXX")
+	_SCAN_CLEANUP_FILES=("$migratable_file" "$non_migratable_file")
 	_save_cleanup_scope
 	trap '_run_cleanups' RETURN
 	push_cleanup "rm -f '$migratable_file' '$non_migratable_file'"
@@ -336,7 +341,7 @@ cmd_migrate() {
 
 			if [[ ${#x86_list[@]} -gt 0 ]]; then
 				print_info "Removing ${#x86_list[@]} x86 packages..."
-				if "$X86_BREW" uninstall --force --ignore-dependencies "${x86_list[@]}" 2>&1 | tail -3; then
+				if "$X86_BREW" uninstall --force --ignore-dependencies "${x86_list[@]}"; then
 					removed_dups=${#x86_list[@]}
 					print_success "Removed ${#x86_list[@]} x86 packages"
 				else

